drivers/pinctrl/pinctrl_silabs_siwx917.c

```diff
@@ -1,25 +1,48 @@
 /*
  * Copyright (c) 2023 Antmicro
+ * Copyright (c) 2024 Silicon Laboratories Inc.
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #define DT_DRV_COMPAT silabs_siwx917_pinctrl
 
 #include <zephyr/drivers/pinctrl.h>
-#include "rsi_rom_egpio.h"
+#include "sl_si91x_peripheral_gpio.h"
 
-static void pinctrl_siwx917_set(bool use_ulp, uint8_t port, uint8_t pin, uint8_t mux)
+#define MODE_COUNT               16
+#define HP_PERIPHERAL_ON_ULP_PIN 6
+
+static inline bool valid_mode(uint8_t mode)
 {
-	/* FIXME: References to EGPIO1 and EGPIO must be replaced by
-	 * DT_REG_ADDR(DT_NODELABEL(pinctl)) (Note Siwx917 has two pinctl instances)
-	 */
-	if (use_ulp) {
-		RSI_EGPIO_UlpPadReceiverEnable(pin);
-		RSI_EGPIO_SetPinMux(EGPIO1, port, pin, mux);
+	return (mode < MODE_COUNT);
+}
+
+static void pinctrl_siwx917_set(uint8_t port, uint8_t pin, uint8_t ulppin, uint8_t mode,
+				uint8_t ulpmode, uint8_t pad)
+{
+	if (pad == 0) {
+		sl_si91x_gpio_enable_host_pad_selection((port << 4) | pin);
+	} else if (pad != 0xFF && pad != 9) {
+		sl_si91x_gpio_enable_pad_selection(pad);
+	}
+
+	/* TODO only enable receiver if input */
+	if (port == SL_GPIO_ULP_PORT) {
+		sl_si91x_gpio_enable_ulp_pad_receiver(ulppin);
 	} else {
-		RSI_EGPIO_HostPadsGpioModeEnable(pin);
-		RSI_EGPIO_PadReceiverEnable(pin);
-		RSI_EGPIO_SetPinMux(EGPIO, port, pin, mux);
+		sl_si91x_gpio_enable_pad_receiver((port << 4) | pin);
+	}
+
+	if (valid_mode(mode)) {
+		GPIO->PIN_CONFIG[((port << 4) | pin)].GPIO_CONFIG_REG_b.MODE = mode;
+	}
+
+	if (valid_mode(ulpmode)) {
+		if (valid_mode(mode) && ulpmode != HP_PERIPHERAL_ON_ULP_PIN) {
```

Comment:
> why selection of ULP peripherals on HP pins doesn't work properly
Regarding this issue, the condition above is suspicious. I believe this condition is never true in `siwx917-pinctrl.h`. I believe `valid_mode(mode)` iff `ulpmode == HP_PERIPHERAL_ON_ULP_PIN`.

Reply:
It was never true because I didn't push the entries that would make it true as part of the draft PR. I'm pushing a change now that adds these, and also makes the change required to make them work (ULP_GPIO MODE must be explicitly set to 0 in addition to the function call).